### ISSUE TITLE
Add npm 2.x.x support

### ISSUE THREAD

**Issue by koistya** [projectkudu/kudu#1427]
### package.json

```json
{
  "engines": {
    "node": "0.10.x",
    "npm": "2.1.x"
  }
}
```

### Error

```
...
remote: No available npm version matches application's version constraint of '2.1.x'. Use
package.json to choose one of the following versions: 1.1.21, 1.1.36, 1.1.37, 1.2.10, 1.2.
18, 1.2.30, 1.3.11, 1.3.21, 1.3.8, 1.4.10, 1.4.23, 1.4.28, 1.4.3, 1.4.9.
```

**Comment by suwatch** [comment]
I will add the latest npm [2.1.17](https://github.com/npm/npm/releases).   This npm is not associated with any node 0.10.x.   The latest node [0.10.35](http://nodejs.org/download/) is still using npm [1.4.28](https://raw.githubusercontent.com/joyent/node/v0.10.35/deps/npm/package.json).

**Comment by suwatch** [comment]
@koistya  do you explicitly want npm 2.1.x?   Should it be 2.0.x?   Even with latest node drop (0.11.14), it is using npm 2.0.0  (https://raw.githubusercontent.com/joyent/node/v0.11.14/deps/npm/package.json).   My point is we should install one that has been officially tested with existing node version.

**Comment by ljharb** [comment]
@suwatch It's always safe to use the latest npm, and `2.1.17` is the latest. `npm` depends on `node`, not the other way around, and every `npm` release is tested extensively with `node`.

**Comment by koistya** [comment]
@suwatch yep, the latest final version of `npm`. What the reason to use a previous version, especially 1.x?

**Comment by davidebbo** [comment]
I think there are two separate topics here that are getting a bit mixed up.

The first is about the **default** npm to use for a given version of Node. Each Node version is 'mapped' to a version of npm. e.g. per [here](https://raw.githubusercontent.com/joyent/node/v0.10.35/deps/npm/package.json), Node `0.10.35` is mapped to npm `1.4.28`. Isn't there a reason for this mapping? If it is always safe to use the latest version of npm regardless of the version of Node, why aren't they pointing to the latest `2.x` npm? Believe me, we'd love it if we only needed to keep one version of npm and be guaranteed that it works for every Node user. But we'd want to be really sure before making that change.

The second topic is about what 2.x version of npm we should make available for those who choose to use it (i.e. non-default). My understanding was that in the Node world, an odd second number meant an 'unstable' version. Hence Node `0.8.x` and `0.10.x` are 'stable', but `0.9.x` and `0.11.x` are not. Now the question is whether this same scheme applies to npm. Looking at this [historical mapping list](http://nodejs.org/dist/npm-versions.txt), it would appear *not* to be the case, as many stable Node builds were mapped to `1.1.x` and `1.3.x` npm. So for the **non-default** npm `2.x` that we make available, it should be fine to go with `2.1.17`.

**Comment by ljharb** [comment]
@davidebbo There's a lot of disagreement about that - basically, the node core team doesn't want to include it simply because the version number is a major bump. However, the npm team themselves has been advocating for it to be fully upgraded in 0.10.x.

Only `node` itself follows the Apache even/odd stable/unstable release pattern - the rest of the node community follows semver, as does npm.

Certainly you're correct about default (`0.10.35` => `1.4.28`) and non-default (`2.1.17`, or really, whatever is the `latest`). You can of course make your own call as to whether you ship with the stock npm, or whether you give your users the latest one.

**Comment by davidebbo** [comment]
Ok, so let's start simple and get `2.1.17` on there as an opt-in version (via package.json). We'll see how that works and we can go from there.

**Comment by suwatch** [comment]
Out of curiosity, is there any feature from npm 2.1.x that you need but wasn't available on older npm?  Or is the requirement for 2.1.x dictated by other package you depend on?

**Comment by koistya** [comment]
With `npm 2.1.x` you can execute scripts even without having them installed globally, also you can pass arguments to npm scripts:

For example, you have `scripts: { "build": "gulp build" }` in your `package.json`

Which you can run with the `--release` flag like this:

`npm run build -- --release`

**Comment by suwatch** [comment]
Could you share any public repo that exercising npm 2.1.x?   We just want to verify that it works once we have added the new npm.

**Comment by suwatch** [comment]
@koistya  We have deployed npm 2.1.17 support to one of our env.  Could you try to push your repo to `https://$npmtest:53wgacddKhvgHdZCz7lpysirj6iinPgtwp8f3Lv2dM6B9dN2egmRi0ArGA7M@npmtest.scm.kudu1.antares-test.windows-int.net/npmtest.git` and see if any issue.

**Comment by davidebbo** [comment]
Couple more things:
- To access Kudu (for debugging and to look at things), go to the URL above, but replace `/npmtest.git` by `basicauth`
- To run the site: http://npmtest.kudu1.antares-test.windows-int.net/

Obviously, this is just a test temporary site, which is why we're giving the creds openly :)

**Comment by suwatch** [comment]
I closed this as 2.1.17 is now available on production.   reopen if otherwise.

**Comment by koistya** [comment]
Cool, great news!

**Comment by davidebbo** [comment]
@koistya: would be great if you could give it a quick try to validate that it works in your scenario. Thanks!

**Comment by koistya** [comment]
It works fine. I noticed that during a deployment it prints what `node.js` version is used, but doesn't tell what `npm` version is used:

```shell
remote: Node.js versions available on the platform are: 0.6.17, 0.6.20, 0.8.2, 0.8.19, 0.8.26...
remote: Selected node.js version 0.10.32. Use package.json file to choose a different version.
```

**Comment by suwatch** [comment]
Right.  It did not print by default.  Our test repo (https://github.com/KuduApps/NodeHelloWithNpm) happened to have custom deployment script that run `%NPM_CMD% --version` explicitly.  Hence we can see it was using what NPM version in the output.
